### PR TITLE
[VL][CELEBORN] Fallback to row-based celeborn shuffle writer

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

If the shuffle exchange is row-based and user enabled the celeborn, we should use vanilla celeborn shuffle writer rather than  fallback to spark shuffle writer. So we can make sure the dra can still work with celeborn.

## How was this patch tested?

manually test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @kerwin-zk thank you

File: gluten-celeborn/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java [hunk 1]
```diff
@@ -54,6 +54,7 @@ public class CelebornShuffleManager implements ShuffleManager {
   private LifecycleManager lifecycleManager;
   private ShuffleClient shuffleClient;
   private volatile ColumnarShuffleManager _columnarShuffleManager;
+  private volatile SparkShuffleManager _vanillaCelenornShuffleManager;
```
Comment (kerwin-zk) [inline gluten-celeborn/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java:57]: _vanillaCelenornShuffleManager -> _vanillaCelebornShuffleManager

File: gluten-celeborn/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java [hunk 2]
```diff
@@ -240,12 +260,17 @@ public <K, V> ShuffleWriter<K, V> getWriter(
...
-        if (ShuffleMode.HASH.equals(celebornConf.shuffleWriterMode())) {
+        if (!ShuffleMode.HASH.equals(celebornConf.shuffleWriterMode())) {
+          throw new UnsupportedOperationException(
+              "Unrecognized shuffle write mode!" + celebornConf.shuffleWriterMode());
+        }
+        if (h.dependency() instanceof ColumnarShuffleDependency) {
+          // columnar-based shuffle
           return new CelebornHashBasedColumnarShuffleWriter<>(
               h, context, celebornConf, client, metrics);
         } else {
```
Comment (kerwin-zk) [inline gluten-celeborn/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java:271]: ditto.

File: gluten-celeborn/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java [hunk 3]
```diff
@@ -160,33 +173,40 @@ private void initializeLifecycleManager() {
...
     if (dependency instanceof ColumnarShuffleDependency) {
-      appUniqueId = SparkUtils.appUniqueId(dependency.rdd().context());
-      initializeLifecycleManager();
-
       if (fallbackPolicyRunner.applyAllFallbackPolicy(
           lifecycleManager, dependency.partitioner().numPartitions())) {
         logger.warn("Fallback to ColumnarShuffleManager!");
         columnarShuffleIds.add(shuffleId);
         return columnarShuffleManager().registerShuffle(shuffleId, dependency);
       } else {
```
Comment (kerwin-zk) [inline gluten-celeborn/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java:202]: Adding concrete types here will make the logic clearer. ```else if (h.dependency() instanceof ShuffleDependency)```, and if none match, an exception is thrown in the else.

File: gluten-celeborn/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java (hunk 2 above)
Comment (ulysses-you) [reply gluten-celeborn/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java:271]: the dependency is awlays instanceof ShuffleDependency.. so it seems the clause is unnecessary

**Review by kerwin-zk — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Pull request merged**